File: ros2_edge_perception/tracker_3d.py
"""
Tier-1 3D Multi-Object Tracking Engine (3D Kalman Filter & Association).

Engineered for Autonomous Mobile Robots (AMRs) and Autonomous Vehicles.
Features:
- Constant-Velocity 3D Kalman Filter: State [x, y, z, vx, vy, vz, sx, sy, sz]
- Track Lifecycle State Machine: Tentative -> Confirmed -> Lost -> Deleted
- Occlusion Handling: Predicts obstacle trajectory across sensor dropouts
- Metric Velocity Vector Estimation: Computes 3D velocity in m/s
- Time-To-Collision (TTC) Risk Evaluation for predictive collision avoidance
- Temporal Association via 3D Euclidean & Bounding Box Overlap
"""

import math
import time
from typing import Dict, List, Optional, Tuple
import numpy as np


class Track3D:
    """Represents a single tracked dynamic 3D obstacle."""

    _next_id = 1

    def __init__(self, detection: dict, timestamp: float):
        self.track_id = Track3D._next_id
        Track3D._next_id += 1

        self.class_name = detection["class_name"]
        self.class_id = detection["class_id"]
        self.score = detection["score"]

        # State vector: [x, y, z, vx, vy, vz, sx, sy, sz]
        self.x = np.zeros(9, dtype=np.float32)
        self.x[0] = detection["x"]
        self.x[1] = detection["y"]
        self.x[2] = detection["z"]
        self.x[3] = 0.0  # vx
        self.x[4] = 0.0  # vy
        self.x[5] = 0.0  # vz
        self.x[6] = detection["size_x"]
        self.x[7] = detection["size_y"]
        self.x[8] = detection["size_z"]

        # State Covariance Matrix P (9x9)
        self.P = np.diag([
            0.1, 0.1, 0.1,    # Positional uncertainty (0.1m)
            1.0, 1.0, 1.0,    # Velocity uncertainty (1.0 m/s)
            0.05, 0.05, 0.05  # Size uncertainty
        ]).astype(np.float32)

        # Process Noise Q
        self.Q_pos = 0.05
        self.Q_vel = 0.5
        self.Q_size = 0.01

        # Measurement Noise R (Position & Size)
        self.R = np.diag([0.05, 0.05, 0.1, 0.05, 0.05, 0.05]).astype(np.float32)

        # Lifecycle management
        self.hits = 1
        self.age = 1
        self.lost_frames = 0
        self.confirmed = False
        self.last_update_time = timestamp

    def predict(self, dt: float):
        """Kalman Filter Prediction Step."""
        dt = max(0.001, min(0.5, dt))

        # State transition matrix F
        F = np.eye(9, dtype=np.float32)
        F[0, 3] = dt
        F[1, 4] = dt
        F[2, 5] = dt

        # Extrapolate state: x = F * x
        self.x = F @ self.x

        # Process noise covariance Q
        Q = np.diag([
            self.Q_pos * dt, self.Q_pos * dt, self.Q_pos * dt,
            self.Q_vel * dt, self.Q_vel * dt, self.Q_vel * dt,
            self.Q_size * dt, self.Q_size * dt, self.Q_size * dt,
        ]).astype(np.float32)

        # Extrapolate covariance: P = F * P * F^T + Q
        self.P = F @ self.P @ F.T + Q

        self.age += 1
        self.lost_frames += 1

    def update(self, detection: dict, timestamp: float):
        """Kalman Filter Measurement Update Step."""
        self.score = detection["score"]

        # Measurement vector z: [x, y, z, sx, sy, sz]
        z = np.array([
            detection["x"], detection["y"], detection["z"],
            detection["size_x"], detection["size_y"], detection["size_z"]
        ], dtype=np.float32)

        # Measurement matrix H (6x9)
        H = np.zeros((6, 9), dtype=np.float32)
        H[0, 0] = 1.0  # x
        H[1, 1] = 1.0  # y
        H[2, 2] = 1.0  # z
        H[3, 6] = 1.0  # sx
        H[4, 7] = 1.0  # sy
        H[5, 8] = 1.0  # sz

        # Innovation: y = z - H*x
        y = z - H @ self.x

        # Innovation covariance: S = H * P * H^T + R
        S = H @ self.P @ H.T + self.R

        # Optimal Kalman Gain: K = P * H^T * S^-1
        K = self.P @ H.T @ np.linalg.inv(S)

        # Updated state: x = x + K*y
        self.x = self.x + K @ y

        # Updated covariance: P = (I - K*H) * P
        I = np.eye(9, dtype=np.float32)
        self.P = (I - K @ H) @ self.P

        self.hits += 1
        self.lost_frames = 0
        if self.hits >= 2:
            self.confirmed = True
        self.last_update_time = timestamp

    def predict_future_trajectory(self, horizon_seconds: float = 2.0, steps: int = 4) -> List[Tuple[float, float, float]]:
        """Forecast future 3D positions for predictive collision avoidance."""
        future_poses = []
        dt_step = horizon_seconds / steps
        for i in range(1, steps + 1):
            t_future = i * dt_step
            fx = self.x[0] + self.x[3] * t_future
            fy = self.x[1] + self.x[4] * t_future
            fz = self.x[2] + self.x[5] * t_future
            future_poses.append((float(fx), float(fy), float(fz)))
        return future_poses

    def compute_ttc(self) -> Optional[float]:
        """
        Compute Time-To-Collision (TTC) in seconds.
        TTC = Distance / Approach_Speed (if moving towards robot, i.e., vz < 0).
        """
        z_dist = self.x[2]
        vz = self.x[5]
        if vz < -0.2 and z_dist > 0.1:  # Approaching at > 0.2 m/s
            return float(z_dist / abs(vz))
        return None

    def compute_orientation(self) -> Tuple[float, float, float, float, float]:
        """
        Compute 3D oriented bounding box heading (yaw) and quaternion.
        Aligned with velocity vector in ground plane (X-Z in camera frame).
        Returns: (yaw, qx, qy, qz, qw)
        """
        vx = self.x[3]
        vz = self.x[5]
        speed_sq = vx * vx + vz * vz
        if speed_sq > 0.04:  # > 0.2 m/s
            yaw = float(math.atan2(vx, vz))
            qx = 0.0
            qy = float(math.sin(yaw / 2.0))
            qz = 0.0
            qw = float(math.cos(yaw / 2.0))
        else:
            yaw = 0.0
            qx = 0.0
            qy = 0.0
            qz = 0.0
            qw = 1.0
        return yaw, qx, qy, qz, qw


class MultiObjectTracker3D:
    """
    Tier-1 3D Multi-Object Tracker.
    Manages track lifecycles, association, and predictive obstacle state estimation.
    """

    def __init__(
        self,
        max_lost_frames: int = 5,
        match_distance_threshold: float = 1.2,  # meters
    ):
        self.max_lost_frames = max_lost_frames
        self.match_dist_thresh = match_distance_threshold
        self.tracks: Dict[int, Track3D] = {}
        self.last_timestamp: Optional[float] = None

    def update(self, detections_3d: List[dict], timestamp: float) -> List[dict]:
        """
        Run predict -> associate -> update lifecycle.
        Returns list of active, confirmed 3D tracks with persistent IDs and velocities.
        """
        # Calculate elapsed dt
        if self.last_timestamp is None:
            dt = 0.033
        else:
            dt = max(0.001, timestamp - self.last_timestamp)
        self.last_timestamp = timestamp

        # 1. Prediction step for all existing tracks
        for track in self.tracks.values():
            track.predict(dt)

        # 2. Association step (Greedy nearest-neighbor matching in 3D Euclidean space)
        track_ids = list(self.tracks.keys())
        matched_tracks = set()
        matched_detections = set()

        if len(track_ids) > 0 and len(detections_3d) > 0:
            # Construct distance matrix
            cost_matrix = np.zeros((len(track_ids), len(detections_3d)), dtype=np.float32)
            for i, tid in enumerate(track_ids):
                track_pos = self.tracks[tid].x[:3]
                for j, det in enumerate(detections_3d):
                    det_pos = np.array([det["x"], det["y"], det["z"]], dtype=np.float32)
                    dist = np.linalg.norm(track_pos - det_pos)
                    # Penalize class mismatch
                    if det["class_name"] != self.tracks[tid].class_name:
                        dist += 10.0
                    cost_matrix[i, j] = dist

            # Try Hungarian assignment, fallback to greedy
            try:
                from scipy.optimize import linear_sum_assignment
                row_ind, col_ind = linear_sum_assignment(cost_matrix)
                for r, c in zip(row_ind, col_ind):
                    if cost_matrix[r, c] <= self.match_dist_thresh:
                        matched_tracks.add(track_ids[r])
                        matched_detections.add(c)
                        self.tracks[track_ids[r]].update(detections_3d[c], timestamp)
            except ImportError:
                # Greedy association
                for i, tid in enumerate(track_ids):
                    best_j = -1
                    min_cost = self.match_dist_thresh
                    for j, det in enumerate(detections_3d):
                        if j not in matched_detections and cost_matrix[i, j] < min_cost:
                            min_cost = cost_matrix[i, j]
                            best_j = j
                    if best_j >= 0:
                        matched_tracks.add(tid)
                        matched_detections.add(best_j)
                        self.tracks[tid].update(detections_3d[best_j], timestamp)

        # 3. Create new tracks for unmatched detections
        for j, det in enumerate(detections_3d):
            if j not in matched_detections:
                new_track = Track3D(det, timestamp)
                self.tracks[new_track.track_id] = new_track

        # 4. Prune dead tracks (lost for > max_lost_frames)
        dead_ids = [
            tid for tid, track in self.tracks.items()
            if track.lost_frames > self.max_lost_frames
        ]
        for tid in dead_ids:
            del self.tracks[tid]

        # 5. Format active tracked objects output
        tracked_output = []
        for track in self.tracks.values():
            if track.confirmed:
                future_trajectory = track.predict_future_trajectory(horizon_seconds=2.0, steps=4)
                ttc = track.compute_ttc()
                yaw, qx, qy, qz, qw = track.compute_orientation()

                tracked_output.append(
                    {
                        "track_id": track.track_id,
                        "class_name": track.class_name,
                        "class_id": track.class_id,
                        "score": track.score,
                        "x": float(track.x[0]),
                        "y": float(track.x[1]),
                        "z": float(track.x[2]),
                        "vx": float(track.x[3]),
                        "vy": float(track.x[4]),
                        "vz": float(track.x[5]),
                        "size_x": float(track.x[6]),
                        "size_y": float(track.x[7]),
                        "size_z": float(track.x[8]),
                        "yaw": yaw,
                        "qx": qx,
                        "qy": qy,
                        "qz": qz,
                        "qw": qw,
                        "future_trajectory": future_trajectory,
                        "ttc": ttc,
                        "lost_frames": track.lost_frames,
                    }
                )

        return tracked_output

File: src/tracker_3d.cpp
#include "ros2_edge_perception/tracker_3d.hpp"
#include <cmath>
#include <algorithm>

namespace ros2_edge_perception {

int Track3D::next_id_ = 1;

Track3D::Track3D(const Detection3DInput& det, double timestamp)
    : track_id_(next_id_++),
      class_name_(det.class_name),
      class_id_(det.class_id),
      score_(det.score),
      last_update_time_(timestamp) {
    x_.setZero();
    x_(0) = det.x;
    x_(1) = det.y;
    x_(2) = det.z;
    x_(3) = 0.0f; // vx
    x_(4) = 0.0f; // vy
    x_(5) = 0.0f; // vz
    x_(6) = det.size_x;
    x_(7) = det.size_y;
    x_(8) = det.size_z;

    P_.setZero();
    P_.diagonal() << 0.1f, 0.1f, 0.1f, 1.0f, 1.0f, 1.0f, 0.05f, 0.05f, 0.05f;
}

void Track3D::predict(double dt) {
    float dt_f = static_cast<float>(std::clamp(dt, 0.001, 0.5));

    // State transition matrix F
    Eigen::Matrix<float, 9, 9> F = Eigen::Matrix<float, 9, 9>::Identity();
    F(0, 3) = dt_f;
    F(1, 4) = dt_f;
    F(2, 5) = dt_f;

    // Extrapolate state: x = F * x
    x_ = F * x_;

    // Process noise covariance Q
    Eigen::Matrix<float, 9, 9> Q = Eigen::Matrix<float, 9, 9>::Zero();
    Q.diagonal() << q_pos_ * dt_f, q_pos_ * dt_f, q_pos_ * dt_f,
                    q_vel_ * dt_f, q_vel_ * dt_f, q_vel_ * dt_f,
                    q_size_ * dt_f, q_size_ * dt_f, q_size_ * dt_f;

    // Extrapolate covariance: P = F * P * F^T + Q
    P_ = F * P_ * F.transpose() + Q;

    age_++;
    lost_frames_++;
}

void Track3D::update(const Detection3DInput& det, double timestamp) {
    score_ = det.score;

    // Measurement vector z: [x, y, z, sx, sy, sz]
    Eigen::Matrix<float, 6, 1> z;
    z << det.x, det.y, det.z, det.size_x, det.size_y, det.size_z;

    // Measurement matrix H (6x9)
    Eigen::Matrix<float, 6, 9> H = Eigen::Matrix<float, 6, 9>::Zero();
    H(0, 0) = 1.0f;
    H(1, 1) = 1.0f;
    H(2, 2) = 1.0f;
    H(3, 6) = 1.0f;
    H(4, 7) = 1.0f;
    H(5, 8) = 1.0f;

    // Measurement noise R
    Eigen::Matrix<float, 6, 6> R = Eigen::Matrix<float, 6, 6>::Zero();
    R.diagonal() << 0.05f, 0.05f, 0.1f, 0.05f, 0.05f, 0.05f;

    // Innovation: y = z - H*x
    Eigen::Matrix<float, 6, 1> y = z - H * x_;

    // Innovation covariance: S = H * P * H^T + R
    Eigen::Matrix<float, 6, 6> S = H * P_ * H.transpose() + R;

    // Kalman Gain: K = P * H^T * S^-1
    Eigen::Matrix<float, 9, 6> K = P_ * H.transpose() * S.inverse();

    // Updated state: x = x + K*y
    x_ = x_ + K * y;

    // Updated covariance: P = (I - K*H) * P
    Eigen::Matrix<float, 9, 9> I = Eigen::Matrix<float, 9, 9>::Identity();
    P_ = (I - K * H) * P_;

    hits_++;
    lost_frames_ = 0;
    if (hits_ >= 2) {
        confirmed_ = true;
    }
    last_update_time_ = timestamp;
}

std::vector<std::tuple<float, float, float>> Track3D::predict_future_trajectory(float horizon, int steps) const {
    std::vector<std::tuple<float, float, float>> trajectory;
    float dt_step = horizon / static_cast<float>(steps);
    for (int i = 1; i <= steps; ++i) {
        float t_future = i * dt_step;
        float fx = x_(0) + x_(3) * t_future;
        float fy = x_(1) + x_(4) * t_future;
        float fz = x_(2) + x_(5) * t_future;
        trajectory.emplace_back(fx, fy, fz);
    }
    return trajectory;
}

std::optional<float> Track3D::compute_ttc() const {
    float z_dist = x_(2);
    float vz = x_(5);
    if (vz < -0.2f && z_dist > 0.1f) {
        return std::abs(z_dist / vz);
    }
    return std::nullopt;
}

void Track3D::compute_orientation(float& yaw, float& qx, float& qy, float& qz, float& qw) const {
    float vx = x_(3);
    float vz = x_(5);
    float speed_sq = vx * vx + vz * vz;

    // If speed > 0.2 m/s, align heading with velocity vector in X-Z ground plane
    if (speed_sq > 0.04f) {
        yaw = std::atan2(vx, vz); // Angle in camera coordinates relative to forward Z
        qx = 0.0f;
        qy = std::sin(yaw / 2.0f);
        qz = 0.0f;
        qw = std::cos(yaw / 2.0f);
    } else {
        yaw = 0.0f;
        qx = 0.0f;
        qy = 0.0f;
        qz = 0.0f;
        qw = 1.0f;
    }
}

MultiObjectTracker3D::MultiObjectTracker3D(int max_lost_frames, float match_distance_thresh)
    : max_lost_frames_(max_lost_frames),
      match_distance_thresh_(match_distance_thresh) {}

std::vector<TrackedObject3D> MultiObjectTracker3D::update(
    const std::vector<Detection3DInput>& detections, double timestamp) {
    double dt = last_timestamp_.has_value() ? std::max(0.001, timestamp - last_timestamp_.value()) : 0.033;
    last_timestamp_ = timestamp;

    // 1. Prediction step
    for (auto& [id, track] : tracks_) {
        track->predict(dt);
    }

    // 2. Greedy 3D Euclidean Association
    std::vector<int> track_ids;
    for (const auto& [id, track] : tracks_) {
        track_ids.push_back(id);
    }

    std::vector<bool> matched_detections(detections.size(), false);
    std::vector<bool> matched_tracks(track_ids.size(), false);

    for (size_t i = 0; i < track_ids.size(); ++i) {
        int tid = track_ids[i];
        const auto& track_state = tracks_[tid]->get_state();
        Eigen::Vector3f track_pos(track_state(0), track_state(1), track_state(2));

        int best_j = -1;
        float min_dist = match_distance_thresh_;

        for (size_t j = 0; j < detections.size(); ++j) {
            if (matched_detections[j]) continue;

            Eigen::Vector3f det_pos(detections[j].x, detections[j].y, detections[j].z);
            float dist = (track_pos - det_pos).norm();

            if (detections[j].class_name != tracks_[tid]->get_class_name()) {
                dist += 10.0f; // Class mismatch penalty
            }

            if (dist < min_dist) {
                min_dist = dist;
                best_j = static_cast<int>(j);
            }
        }

        if (best_j >= 0) {
            matched_tracks[i] = true;
            matched_detections[best_j] = true;
            tracks_[tid]->update(detections[best_j], timestamp);
        }
    }

    // 3. Create new tracks for unmatched detections
    for (size_t j = 0; j < detections.size(); ++j) {
        if (!matched_detections[j]) {
            auto new_track = std::make_unique<Track3D>(detections[j], timestamp);
            tracks_[new_track->get_id()] = std::move(new_track);
        }
    }

    // 4. Prune stale tracks
    for (auto it = tracks_.begin(); it != tracks_.end(); ) {
        if (it->second->get_lost_frames() > max_lost_frames_) {
            it = tracks_.erase(it);
        } else {
            ++it;
        }
    }

    // 5. Output confirmed tracks
    std::vector<TrackedObject3D> active_tracks;
    for (const auto& [id, track] : tracks_) {
        if (track->is_confirmed()) {
            const auto& s = track->get_state();
            TrackedObject3D obj;
            obj.track_id = track->get_id();
            obj.class_name = track->get_class_name();
            obj.class_id = track->get_class_id();
            obj.score = track->get_score();
            obj.x = s(0);
            obj.y = s(1);
            obj.z = s(2);
            obj.vx = s(3);
            obj.vy = s(4);
            obj.vz = s(5);
            obj.size_x = s(6);
            obj.size_y = s(7);
            obj.size_z = s(8);
            track->compute_orientation(obj.yaw, obj.qx, obj.qy, obj.qz, obj.qw);
            obj.future_trajectory = track->predict_future_trajectory(2.0f, 4);
            obj.ttc = track->compute_ttc();
            obj.lost_frames = track->get_lost_frames();
            active_tracks.push_back(obj);
        }
    }

    return active_tracks;
}

} // namespace ros2_edge_perception
